### PR TITLE
[LF] resurect lf.data.Decimal

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
but make it deprecated

### CODE REVIEW

**Review by meiersi-da — APPROVED** [review]
Nice! Thanks!

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/NumericModule.scala [hunk 1]
```diff
@@ -225,6 +225,21 @@ abstract class NumericModule {
   def assertFromString(s: String): Numeric =
     assertRight(fromString(s))
 
+  /** Given a string representation of a decimal and a scale s returns the corresponding `Numeric s`.
+    * If the input does not match (where i = 38-s)
+    *   `-?([1-9]\d{1,i}|0).([1-9]{s})`
+    * or if the result of the conversion cannot be mapped into a numeric without loss of precision
+    * returns an error message instead.
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/NumericModule.scala:232]:
```suggestion
    * or if the result of the conversion cannot be mapped into a numeric of scale `s` without loss 
    * of precision returns an error message instead.
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  @deprecated("Use Numeric.fromBigDecimal(Scale.assertFromInt(10), _)", since = "3.0.0")
+  def fromBigDecimal(x: BigDec): Either[String, data.Numeric.Numeric] =
+    Numeric.fromBigDecimal(scale, x)
+
+  private val hasExpectedFormat =
+    """[+-]?\d{1,28}(\.\d{1,10})?""".r.pattern
+
+  @deprecated(
+    """Use Numeric.fromString(Scale.assertFromInt(10), _). Note this alternative is stricter as it requires the string to match "`-?[1-9]\d{1,28}.([1-9]{10})`"""",
```
Comment (meiersi-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala:32]: @remyhaemmerle-da : what is the reason for the more strict matching?

I've just run into that causing a test-failure.

Comment (meiersi-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala:32]: That said. We've decided to become more strict in our decoding as well. You can ignore this comment.

Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala:32]: The reason is that I do not want to write and maintain a new parser. 
In `Numeric.fromString(_, _)` reuses `Numeric.fromString(_)` that uses the number of digits on the right and side of the '.' to determine the scale.

If you want something more lenient, I suggest you use BigDecimal parser and convert the result with `Numeric.fromBigDecimal` 

```
def convert(s: String) =  Numeric.assertFromBigDecimal(Numeric.Scale.assertFromInt(10), BigDecimal(s))
```

Comment (meiersi-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala:32]: ack. thx. I ended up going that path while keeping the format restriction we've had for Decimal.

**Pull request merged**